fix pimd/langevin: support running for group

Kehan-Cai-nanako opened the pull request:
**Summary**

The current code of `fix ID group-ID pimd/langevin keyword value` runs for all atoms even if this fix style is set for a target group of atoms in the input file, e.g, `fix pimd/langevin groupA`.

The code of this PR is designed to support running `fix pimd/langevin` for a target group of atoms, by masking atoms appropriately in `fix_pimd_langevin.cpp`, specifically for some functions of integration and computation of physical observables. 

**Author(s)**

Kehan Cai, Princeton University

Yifan Li @Yi-FanLi, Princeton University

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No changes in the pull request will break backward compatibility for inputs.

**Implementation Notes**

When performing simulations with `fix ID group-ID pimd/langevin keyword value`, the group-ID should be set as the target group of atoms. 

The correctness was verified by comparing results of the 1 replica/bead PIMD simulations with those of classical MD simulations. Besides, for multiple-replicas/beads situations, temperatures, pressures, virials and energies were checked if they were consistent according to formulas and mutual relations.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->
[example_PIMD_PR.zip](https://github.com/user-attachments/files/23173441/example_PIMD_PR.zip)

Kehan-Cai-nanako committed:
Add mask loop for atoms

Kehan-Cai-nanako committed:
Add mask loop for atoms and change natoms to group count

Kehan-Cai-nanako committed:
Remove some masks

akohlmey commented on src/REPLICA/fix_pimd_langevin.cpp:
```diff
@@ -514,10 +514,15 @@ void FixPIMDLangevin::init()
   imageint *image = atom->image;
   if (mapflag) {
-    for (int i = 0; i < nlocal; i++) domain->unmap(x[i], image[i]);
+    for (int i = 0; i < nlocal; i++) {
```
There is no actual change in this function, the original code should be restored and the comments removed.

akohlmey commented on src/REPLICA/fix_pimd_langevin.cpp:
```diff
@@ -550,7 +555,11 @@ void FixPIMDLangevin::setup(int vflag)
   }
   if (mapflag) {
-    for (int i = 0; i < nlocal; i++) domain->unmap_inv(x[i], image[i]);
+    for (int i = 0; i < nlocal; i++) {
```
There is no actual change in this function, the original code should be restored and the comments removed.

akohlmey commented on src/REPLICA/fix_pimd_langevin.cpp:
```diff
@@ -565,10 +574,15 @@ void FixPIMDLangevin::setup(int vflag)
   imageint *image = atom->image;
   if (mapflag) {
-    for (int i = 0; i < nlocal; i++) domain->unmap(x[i], image[i]);
+    for (int i = 0; i < nlocal; i++) {
```
There is no actual change in this function, the original code should be restored and the comments removed.

akohlmey commented on src/REPLICA/fix_pimd_langevin.cpp:
```diff
@@ -663,7 +677,11 @@ void FixPIMDLangevin::initial_integrate(int /*vflag*/)
 
   if (mapflag) {
-    for (int i = 0; i < nlocal; i++) { domain->unmap_inv(x[i], image[i]); }
+    for (int i = 0; i < nlocal; i++) {
```
Same as above

akohlmey commented on src/REPLICA/fix_pimd_langevin.cpp:
```diff
@@ -710,17 +729,25 @@ void FixPIMDLangevin::post_force(int /*flag*/)
-    x_unwrap[i][0] = x[i][0];
-    x_unwrap[i][1] = x[i][1];
-    x_unwrap[i][2] = x[i][2];
+    // if (mask[i] & groupbit) {
```
There is no actual change in this function, the original code should be restored and the comments removed.

akohlmey reviewed (changes requested):
I don't understand the purpose of submitting a pull request with modifications, most of which are then commented out. Why are those changes commented? This seems counter-intuitive given the original description of the pull request. Please explain.

Also, there should be a simple/small example provided that clearly demonstrates the changed functionality.

Kehan-Cai-nanako committed:
Uncomment mask atoms in equation of motion integration steps

Kehan-Cai-nanako committed:
Clean up comments for pull request

Kehan-Cai-nanako commented:
Thank you for reviewing the code.

Initially, I added the changes related to `mask & groupbit` in all places where I believed they were necessary. However, subsequent tests indicated that some of these changes were not needed, so I temporarily commented them out. Specifically, the changes related to mask should not be done in functions related to communications and transformations among processes. I have now removed those commented lines from the code.

I have also revised the summary section of this PR to make it more concise and easier to follow, removing unnecessary details that might have caused confusion. In short, this PR enables the use of `fix pimd/langevin` for a specific group of atoms rather than for all atoms, aligning with the functionality already established for other fix styles.

I hope this clarification is clear. Please let me know if any additional information or adjustments are needed.

akohlmey commented:
> I hope this clarification is clear. Please let me know if any additional information or adjustments are needed.

As I mentioned before, there should be a simple/minimal example that demonstrates the impact of the changes and that can be used for validation.

akohlmey committed:
Merge branch 'develop' into feature/fix_pimd_langevin_mask_atoms

Kehan-Cai-nanako committed:
Add an example for PIMD group functionality

Kehan-Cai-nanako committed:
Merge remote-tracking branch 'origin/feature/fix_pimd_langevin_mask_atoms' into feature/fix_pimd_langevin_mask_atoms

Kehan-Cai-nanako commented:
> > I hope this clarification is clear. Please let me know if any additional information or adjustments are needed.
> 
> As I mentioned before, there should be a simple/minimal example that demonstrates the impact of the changes and that can be used for validation.

I have added an example for the group functionality of `fix pimd/langevin` in the [examples/PACKAGES/pimd/langevin_reduecd_units_group](https://github.com/Kehan-Cai-nanako/lammps/tree/feature/fix_pimd_langevin_mask_atoms/examples/PACKAGES/pimd/langevin_reduced_units_group) folder. The example is also attached in the summary section for your convenience.

megmcca commented:
@Kehan-Cai-nanako thanks for the example including the run command you've used. i can confirm that the new code compiles and the example runs to completion with 4x1 processors on my local workstation. 

to complete your example, could you also upload output files similar to the other `examples/PACKAGES/pimd/langevin*` directories (the four per-proc log files)?

Kehan-Cai-nanako committed:
Add output log files

Kehan-Cai-nanako commented:
> @Kehan-Cai-nanako thanks for the example including the run command you've used. i can confirm that the new code compiles and the example runs to completion with 4x1 processors on my local workstation.
> 
> to complete your example, could you also upload output files similar to the other `examples/PACKAGES/pimd/langevin*` directories (the four per-proc log files)?

Thanks for testing the code! I have uploaded output log files to the corresponding example folder.

megmcca committed:
rename with date format

megmcca committed:
Merge branch 'develop' into feature/fix_pimd_langevin_mask_atoms

megmcca reviewed (approved):
updated log file names, i think it's ready to go! approved.

akohlmey commented:
> I hope this clarification is clear. Please let me know if any additional information or adjustments are needed.

Sorry, but no. If the fix is applied only on a specific group of atoms, why should it touch *any* data that belongs to *other* atoms except for data that is maintained internally and will not lead to any changes of the non-group atoms? The fact that it does not make a difference for your specific test example is not a sufficient argument.

Please note that those other atoms my be moved by a different fix (e.g. fix move or fix nve).

Kehan-Cai-nanako commented:
> > I hope this clarification is clear. Please let me know if any additional information or adjustments are needed.
> 
> Sorry, but no. If the fix is applied only on a specific group of atoms, why should it touch _any_ data that belongs to _other_ atoms except for data that is maintained internally and will not lead to any changes of the non-group atoms? The fact that it does not make a difference for your specific test example is not a sufficient argument.
> 
> Please note that those other atoms my be moved by a different fix (e.g. fix move or fix nve).

The data of the non-group atoms are accessed only within the routines responsible for communication and normal-mode transformations. The communication operations, i.e., `inter_replica_communications()`, do not alter any data of the non-group atoms. Regarding the normal-mode transformations, the operations on the non-group atoms effectively involve applying a transformation and its exact inverse transformation to their coordinates, which leaves these atoms unchanged. In summary, no changes are made to the non-group atoms, thereby preserving the correctness of the molecular dynamics simulations.

The only minor issue is that the current implementation introduces some redundant communication and transformation steps for the non-group atoms. This design choice was made to simplify the implementation and maintain code clarity. Specifically, during the communication of atomic information across replicas, atoms are sorted by their tags rather than by indices, which makes it challenging to exclude the non-group atoms efficiently. Implementing such an exclusion would require additional mapping operations, thereby increasing computational overhead and resource usage. Therefore, we chose the current approach as a balanced solution, considering the trade-off between the two routines.
